src/units.rs

```diff
@@ -0,0 +1,385 @@
+use crate::{
+    member::NotificationOut, Data, Hash, KeyBox, NodeCount, NodeIndex, NodeMap, Round, SessionId,
+};
+use codec::{Decode, Encode};
+use log::{debug, error};
+use std::{collections::HashMap, hash::Hash as StdHash};
+
+// TODO: need to make sure we never accept units of round > MAX_ROUND
+pub(crate) const MAX_ROUND: usize = 5000;
+
+#[derive(Debug, Clone, Encode, Decode)]
+pub(crate) struct FullUnit<H: Hash, D: Data> {
+    pub(crate) inner: PreUnit<H>,
+    pub(crate) data: D,
+    pub(crate) session_id: SessionId,
+}
+
+#[derive(Debug, Clone, Encode, Decode)]
+pub(crate) struct SignedUnit<H: Hash, D: Data, Signature: Clone + Encode + Decode> {
+    pub(crate) unit: FullUnit<H, D>,
+    pub(crate) signature: Signature,
+    pub(crate) id: NodeIndex,
```

Comment:
We have a small redundancy here: the node index is stored in the `id` and implicitly in `signature`. Maybe we could remove the `id` field and adding a constraint that Signature should implement the `MyIndex` trait?

Reply:
Good idea :) Changed.

Reply:
1) remove id from the struct and provide id() as a method that reads it off the unit
2) I'm against Signature carrying the index. Unless there is a workaround, then this index must be serialized and sent over network, which there is no need for. To fix this, KeyBox verify() function should also take a NodeIndex indicating who signed.

Reply:
Concerning 2. I think it would be nice if `Signature` indicated whose signature it represents (as in this case by containing the id), but I also agree that unnecessarily sending this data over the network is at least annoying and possibly impactingly inefficient. Maybe this could be worked around in the `Signed<T>` struct suggested by @michalseweryn, by implementing a specific version of serialization there and requiring that the internal type has an index in some sense?

Reply:
It seems that we have a discussion about a bigger refactor of how we do signing spread around at least three comments – maybe let's leave these issues alone for this PR and I can try cooking up some design for such a refactor over the next couple of days. The design would be in a gdoc to make commenting on specific properties easier.